test/suites/cloud-init.sh

```diff
@@ -27,4 +27,79 @@ test_cloud_init() {
   [ -n "${ID6}" ] && [ "${ID6}" != "${ID5}" ]
 
   lxc delete -f c1 c2
+
+  ensure_import_testimage
+  fingerprint="$(lxc image info testimage | awk '/^Fingerprint:/ {print $2}')"
+
+  # Ensure testimage is not set as cached.
+  lxd sql global "UPDATE images SET cached=0 WHERE fingerprint=\"${fingerprint}\""
+
+  (
+    cd devlxd-client || return
+    # Use -buildvcs=false here to prevent git complaining about untrusted directory when tests are run as root.
+    CGO_ENABLED=0 go build -tags netgo -v -buildvcs=false ./...
+  )
+
+  lxc launch testimage devlxd
+  lxc file push --mode 0755 "devlxd-client/devlxd-client" devlxd/bin/
+
+  # Unknown cloud-init formats should be passed to the instance unmodified
+  lxc config set devlxd cloud-init.user-data="invalid-yaml"
+  [ "$(lxc exec devlxd -- devlxd-client cloud-init user-data)" = "invalid-yaml" ]
+  lxc config unset devlxd cloud-init.user-data
+
+  # With SSH keys configured, unknown cloud-init formats should be passed to the instance unmodified
+  lxc config set devlxd cloud-init.user-data="invalid-yaml"
+  lxc config set devlxd cloud-init.ssh-keys.mykey="root:gh:user1"
+  [ "$(lxc exec devlxd -- devlxd-client cloud-init user-data)" = "invalid-yaml" ]
+  lxc config unset devlxd cloud-init.ssh-keys.mykey
+  lxc config unset devlxd cloud-init.user-data
+
+  # Configured ssh-keys in user-data should not replace vendor-data with users provided
+  CLOUD_INIT="$(cat <<EOF
+#cloud-config
+users:
+  - name: root
+    ssh-import-id: gh:user2
+EOF
+  )"
+
+  lxc config set devlxd cloud-init.vendor-data="${CLOUD_INIT}"
+  lxc config set devlxd cloud-init.ssh-keys.mykey="root:gh:user1"
+  lxc exec devlxd -- devlxd-client cloud-init vendor-data | grep "gh:user1"
+  lxc exec devlxd -- devlxd-client cloud-init vendor-data | grep "gh:user2"
+  [ "$(lxc exec devlxd -- devlxd-client cloud-init user-data)" = "not found" ]
+  lxc config unset devlxd cloud-init.ssh-keys.mykey
+  lxc config unset devlxd cloud-init.vendor-data
+
+  # Valid format should work with ssh keys embedded
+  CLOUD_INIT="$(cat << EOF
+## template: jinja
+#cloud-config
+runcmd:
+  - echo {{v1.local_hostname}} > /var/tmp/runcmd_output
+  - echo {{merged_system_cfg._doc}} >> /var/tmp/runcmd_output
+EOF
+  )"
+  lxc config set devlxd cloud-init.user-data="${CLOUD_INIT}"
+  lxc config set devlxd cloud-init.ssh-keys.mykey="root:gh:user1"
+  EXPECTED_CLOUD_INIT="$(cat << EOF
+## template: jinja
+#cloud-config
+runcmd:
+- echo {{v1.local_hostname}} > /var/tmp/runcmd_output
+- echo {{merged_system_cfg._doc}} >> /var/tmp/runcmd_output
+users:
+- name: root
+  ssh-import-id:
+  - gh:user1 #lxd:cloud-init.ssh-keys
+  ssh_import_id:
+  - gh:user1 #lxd:cloud-init.ssh-keys
+EOF
+  )"
+  lxc exec devlxd -- devlxd-client cloud-init user-data
+  [ "$(lxc exec devlxd -- devlxd-client cloud-init user-data)" = "${EXPECTED_CLOUD_INIT}" ]
+
+  lxc stop devlxd
```

Comment:
I'd skip the `lxc stop` here as it's quicker to `lxc delete -f` it... we tend to not be nice with test instances ;)